README.md
[![npm version](https://badge.fury.io/js/ethereum-smart-contracts.svg)](https://badge.fury.io/js/ethereum-smart-contracts)
[![dependencies Status](https://david-dm.org/alesanro/ethereum-smart-contracts.svg)](https://david-dm.org/alesanro/ethereum-smart-contracts.svg)
[![devDependencies Status](https://david-dm.org/alesanro/ethereum-smart-contracts/dev-status.svg)](https://david-dm.org/alesanro/ethereum-smart-contracts?type=dev)
# Common information

This package provides an ability to access and use basic smart contracts without any need to compile them manually.

### List of provided smart contract interfaces:
- ERC20
- _to be continued..._

### List of provided tokens:
- MassTransfer ERC20 compatible token


## Usage 

First of all, import a module
```javascript
const ContractsContext = require("ethereum-smart-contracts")
```

Then there might be two ways:
- load it with node uri
```javascript
const context = ContractsContext.withUri("https://mainnet.infura.io/", options)
```
- or inject `web3` instance directly
```javascript
const context = new ContractsContext(web3, options)
```

> `options` parameter should have the next layout: [see parameter `options` structure](https://web3js.readthedocs.io/en/1.0/web3-eth-contract.html#parameters). `from`, `gas` and `gasPrice` are needed to be provided, `data` is optional. 


After initial setup contracts are available to use them.
For example, to load any `ERC20`-token (let use EOS token address) you need to do
```javascript
const tokenAddress = "0x7b39940dbac110f1227d37c395675def270afcd7"
const erc20Token = context.getERC20TokenAt(tokenAddress)
```

```javascript
// to get token's total supply
const totalSupply = await erc20Token.methods.totalSupply().call()
console.log(`#> totalSupply = ${totalSupply}`)
```

```javascript
// or if you want to get user's token balance
const userAddress = "0x00000000000000000000000000000000000000b1"
const balanceOf = await erc20Token.methods.balanceOf(userAddress).call()
console.log(`#> user balance (${userAddress}) = ${balanceOf}`)
```

## Examples

Some example of how to use smart contracts you can find in [examples](https://github.com/alesanro/ethereum-smart-contracts/tree/master/examples) directory

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.0.3"></a>
## [0.0.3](https://github.com/alesanro/ethereum-smart-contracts/compare/v0.0.2-beta.1...v0.0.3) (2018-05-28)


index.js
const Web3 = require("web3")
const contract = require("truffle-contract")

const ERC20Artifact = require("./build/contracts/ERC20")
const BasicTokenArtifact = require("./build/contracts/BasicToken")
const MassTransferERC20Artifact = require("./build/contracts/MassTransferERC20")

const getContractAt = (web3, contractAbi, address, { from, gas, gasPrice, }) => {
	let contract
	try { // try for web3 version 1.*
		contract = new web3.eth.Contract(contractAbi, address, { 
			from: from, 
			gas: gas, 
			gasPrice: gasPrice, 
		})
	}
	catch (e) { // otherwise try for web3 version 0.2.*
		const Scheme = web3.eth.contract(contractAbi)
		contract = Scheme.at(address)	
	}

	return contract
}

/**
 * @class Context creates interface for connecting with 
 * basic smart contracts (ERC20 compatible and other)
 * @param {Web3} web3 web3.js instance
 * @param {object} options options for sending txs from contracts
 */
function Context(web3, options = null) {
	const self = this
	
	this.web3 = web3
	this.options = options || { 
		from: web3.eth.coinbase, 
		gas: 300000, 
		gasPrice: 1
	}

	this.contracts = {
		ERC20: ERC20Artifact,
		BasicToken: BasicTokenArtifact,
		MassTransferERC20: MassTransferERC20Artifact,
	}

	this.getContractAt = (contractAbi, address, options = self.options) => {
		return getContractAt(self.web3, contractAbi, address, options)
	}

	this.getERC20TokenAt = (address, options = self.options) => {
		return getContractAt(self.web3, ERC20Artifact.abi, address, options)
	}
	
	this.getMassTransferERC20TokenAt = (address, options = self.options) => {		
		return getContractAt(self.web3, MassTransferERC20Artifact.abi, address, options)
	}
}

module.exports = Context
module.exports.withUri = (uri, options) => {
	const web3 = new Web3(new Web3.providers.HttpProvider(uri))

	return new Context(web3, options)
}


truffle.js
var HDWalletProvider = require("truffle-hdwallet-provider") /* eslint global-require: 0 */

function getWallet() {
	try {
		return require('fs').readFileSync("./wallet.json", "utf8").trim()
	}
	catch (err) {
		return ""
	}
}

function getPassword() {
	try {
		return JSON.parse(require('fs').readFileSync("./secret.json", "utf8").trim())["password"]
	}
	catch (err) {
		return ""
	}
}

module.exports = {
	// See <http://truffleframework.com/docs/advanced/configuration>
	// to customize your Truffle configuration!

	networks: {
		"private": {
			network_uri_test: "http://127.0.0.1:8540",
			provider: new HDWalletProvider.WalletProvider(getWallet(), getPassword(), "http://127.0.0.1:8540"),
			network_id: 74565,
			gasPrice: 0,
			gas: 4700000,
		},
		test: {
			host: "localhost",
			port: "8545",
			network_id: "*",
			gas: 4700000,
		},
	},
	solc: {
		optimizer: {
			enabled: true,
			runs: 200,
		},
	},
}


examples/fetch-tokens.js
const ContractsContext = require("../index")

module.exports = async function () {
	const uri = "https://mainnet.infura.io/"
	const options = {
		from: "0x4a2d3fc1587494ca2ca9cdeb457cd94be5d96a61",
		gas: 3000000,
		gasPrice: 20
	}
	
	const tokenAddresses = [
		"0x7b39940dbac110f1227d37c395675def270afcd7", // EOSG
		"0x6531f133e6deebe7f2dce5a0441aa7ef330b4e53", // TIME
	]

	const context = ContractsContext.withUri(uri, options)
	
	const tokens = []
	for (var tokenAddress of tokenAddresses) {
		const token = context.getERC20TokenAt(tokenAddress)
		tokens.push(token)
		console.log(`
		##################
		## Token '${await token.methods.symbol().call()}':
		## - name: ${await token.methods.name().call()}
		## - totalSupply: ${await token.methods.totalSupply().call()}
		## - decimals: ${await token.methods.decimals().call()}
		`)
	}
}



examples/mass-transfer-tokens.js
const ContractsContext = require("../index")

module.exports = async function () {
	let context
	let options
	let tokenAddress

	/// First setup: check token presence and setup web3 and the context
	if (web3) {
		const mandatoryContract = artifacts.require("Migrations")
		options = mandatoryContract.defaults()

		context = new ContractsContext(web3, options)
		tokenAddress = (await artifacts.require("BasicToken").deployed()).address
	} else { /// here could be any custom parameters that are needed
		const truffleConfig = require("../truffle")
		const network = truffleConfig.networks["private"]
		const uri = network.network_uri_test
		const tokenHolder = network.provider.address
		options = {
			from: tokenHolder,
			gas: network.gas,
			gasPrice: network.gasPrice,
		}
		context = ContractsContext.withUri(uri, options)
		return
	}

	function SequentialAddressGenerator(seed = 1) {
		seedNumber = seed;

		this.getNextRecepient = () => {
			return {
				address: "0x" + (seedNumber++).toString().padStart(40, "0"), // for web3#v0.2.*
				// address: web3.utils.randomHex(20), // for web3#v1.*
				value: web3.toBigNumber(Math.floor(Math.random()*10000) + 1000), // for web3#v0.2.*
				// value: web3.toBN(Math.floor(Math.random()*10000) + 1000), // for web3#v1.*
			}
		}
	}
	
	/// Create address generator to produce a list of addresses and amounts of tokens
	/// that are supposed to be transferred to them
	const addressGenerator = new SequentialAddressGenerator(1)
	const numberOfRecepients = 10 /// NOTE: could be any positive number
	const recepients = Array.apply(null, {length: numberOfRecepients})
		.map(() => addressGenerator.getNextRecepient())
		.reduce((prev, el) => {
			prev.addresses.push(el.address)
			prev.values.push(el.value)
			return prev
		}, { addresses: [], values: [], })
	
	/// Check if the token address is correct or setup our own token address
	tokenAddress = tokenAddress || "0x" // TODO: mass-transfer compatible token	

	/// Get contract for a provided address and ready to call functions
	const token = context.getMassTransferERC20TokenAt(tokenAddress)

	/// Transfer tokens from tokenHolder to recepients.
	/// Amount of immediate transfers for current options (gasLimit, gasPrice and so on)
	/// could be fetched by using token.massTransfer.call approach. Returned result will be
	/// array of two items [ resultCode, amountOfSuccessfulTransfers, ]
	const tx = await token.massTransfer(recepients.addresses, recepients.values, { 
		from: options.from, 
		gas: options.gas,
		gasPrice: options.gasPrice,
	})
}



examples/example1.js
const SmartContracts = require("../index")

module.exports = async function () {
	const uri = "https://mainnet.infura.io/"
	const options = {
		from: "0x4a2d3fc1587494ca2ca9cdeb457cd94be5d96a61",
		gas: 3000000,
		gasPrice: 20
	}
	
	const tokenAddresses = [
		"0x7b39940dbac110f1227d37c395675def270afcd7", // EOSG
		"0x6531f133e6deebe7f2dce5a0441aa7ef330b4e53", // TIME
	]

	const context = SmartContracts(uri)
	const web3 = context.web3
	
	const tokens = []
	for (var tokenAddress of tokenAddresses) {
		const token = SmartContracts.getERC20TokenAt(web3, tokenAddress, options)
		tokens.push(token)
		console.log(`
		##################
		## Token '${await token.methods.symbol().call()}':
		## - name: ${await token.methods.name().call()}
		## - totalSupply: ${await token.methods.totalSupply().call()}
		## - decimals: ${await token.methods.decimals().call()}
		`)
	}
}

